drivers/crypto/crypto_intel_sha.c

```diff
@@ -0,0 +1,334 @@
+	    (IS_ALIGNED(src, sizeof(uint32_t)))) {
+		len /= sizeof(uint32_t);
+		for (counter = 0; counter != len; ++counter) {
+			((uint32_t *)dst)[counter] = ((uint32_t *)src)[counter];
+		}
+
+		err = 0;
+	}
+
+	return err;
+}
+
+/* ! Perform SHA computation over requested region. */
+static int intel_sha_device_run(struct device *dev, const void *buf_in, size_t buf_in_size,
+				size_t max_buff_len, uint32_t state)
+{
+	int err;
+	struct sha_container *const self = dev->data;
+	union sha_state state_u = { .full = state };
+	/* align to OWORD */
+	const size_t aligned_buff_size = ROUND_UP(buf_in_size, 0x10);
+
+	err = intel_sha_set_ctl_enable(self, 0);
+	if (err)
+		return err;
+
+	/* set processing element disable */
+	self->dfsha->pibcs.part.peen = 0;
+	/* set pib base addr */
+	self->dfsha->pibba.full = (uint32_t)buf_in;
+
+	if (max_buff_len < aligned_buff_size)
+		return -EINVAL;
+
+	self->dfsha->pibs.full = aligned_buff_size;
+	/* enable interrupt */
+	self->dfsha->pibcs.part.bscie = 1;
+	self->dfsha->pibcs.part.teie = 0;
+	/* set processing element enable */
+	self->dfsha->pibcs.part.peen = 1;
+
+	if (self->dfsha->shactl.part.en)
+		return -EINVAL; /* already enabled */
+
+	self->dfsha->shactl.part.hrsm = state_u.part.hrsm;
+
+	/* set initial values if resuming */
+	if (state_u.part.hrsm) {
+		err = intel_sha_set_resume_length_dw0(self, self->dfsha->shaaldw0.full);
+		if (err)
+			return err;
+		err = intel_sha_set_resume_length_dw1(self, self->dfsha->shaaldw1.full);
+		if (err)
+			return err;
+		err = intel_sha_regs_cpy((void *)self->dfsha->initial_vector,
+					 (void *)self->dfsha->sha_result,
+					 sizeof(self->dfsha->initial_vector));
+		if (err)
+			return err;
+	}
+
+	/* set ctl hash first middle */
+	if (self->dfsha->shactl.part.en)
+		return -EINVAL; /* already enabled */
+
+	self->dfsha->shactl.part.hfm = state_u.part.state;
+
+	/* increment pointer */
+	self->dfsha->pibfpi.full = buf_in_size;
+
+	err = intel_sha_set_ctl_enable(self, 1);
+	if (err)
+		return err;
+
+	err = intel_sha_set_ctl_enable(self, 0);
+
+	return err;
+}
+
+static int intel_sha_copy_hash(struct sha_container *const self, void *dst, size_t len)
+{
+	/* NOTE: generated hash value should be read from the end */
+
+	int err = -EINVAL;
+	uint32_t counter = 0;
+	uint32_t last_idx = 0;
+
+	if ((IS_ALIGNED(len, sizeof(uint32_t))) && (IS_ALIGNED(dst, sizeof(uint32_t)))) {
+		len /= sizeof(uint32_t);
+		counter = 0;
+		/* The index of a last element in the sha result buffer. */
+		last_idx = (sizeof(self->dfsha->sha_result) / sizeof(uint32_t)) - 1;
+
+		for (counter = 0; counter != len; counter++) {
+			((uint32_t *)dst)[counter] =
+				((uint32_t *)self->dfsha->sha_result)[last_idx - counter];
+		}
+
+		err = 0;
+	}
+
+	return err;
+}
+
+static int intel_sha_device_get_hash(struct device *dev, void *buf_out, size_t buf_out_size)
+{
+	int err;
+	struct sha_container *const self = dev->data;
+
+	if (dev == NULL)
+		return -ENODEV;
```

Comment:
```suggestion
	if (dev == NULL)
		return -ENODEV;
```
Unnecessary check. This is an internal function and you have already accessed this memory before calling it.

Reply:
removed